lib/rubygems/specification.rb

```diff
@@ -2789,7 +2789,7 @@ def validate packaging = true
 
     validate_dependencies
 
-    true
+    allow_warnings || @warnings.zero?
```

Comment:
You can use [this test](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_specification.rb#L2733-L2741) as a base for testing `spec.validate true, false`. (#validate should raise.)